presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergFilterPushdown.java

```diff
@@ -0,0 +1,149 @@
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergFilterPushdown
+        extends BaseHiveFilterPushdown
+{
+    protected IcebergTransactionManager icebergTransactionManager;
+
+    public IcebergFilterPushdown(
+            RowExpressionService rowExpressionService,
+            StandardFunctionResolution functionResolution,
+            FunctionMetadataManager functionMetadataManager,
+            IcebergTransactionManager transactionManager)
+    {
+        super(rowExpressionService, functionResolution, functionMetadataManager);
+
+        this.icebergTransactionManager = requireNonNull(transactionManager, "transactionManager is null");
+    }
+
+    @Override
+    protected ConnectorPushdownFilterResult getConnectorPushdownFilterResult(
+            Map<String, ColumnHandle> columnHandles,
+            ConnectorMetadata metadata,
+            ConnectorSession session,
+            TupleDomain<Subfield> domainPredicate,
+            RemainingExpressions remainingExpressions,
+            Set<String> predicateColumnNames,
+            Optional<ConnectorTableLayoutHandle> currentLayoutHandle,
+            ConnectorTableHandle tableHandle,
+            HivePartitionResult hivePartitionResult)
+    {
+        Map<String, IcebergColumnHandle> predicateColumns = predicateColumnNames.stream()
+                .map(columnHandles::get)
+                .map(IcebergColumnHandle.class::cast)
+                .collect(toImmutableMap(IcebergColumnHandle::getName, Functions.identity()));
+
+        Optional<Set<IcebergColumnHandle>> requestedColumns = currentLayoutHandle.map(layout -> ((IcebergTableLayoutHandle) layout).getRequestedColumns()).orElse(Optional.empty());
+
+        return new ConnectorPushdownFilterResult(
+                metadata.getTableLayout(
+                        session,
+                        new IcebergTableLayoutHandle(
+                                domainPredicate,
+                                remainingExpressions.getRemainingExpression(),
+                                predicateColumns,
+                                requestedColumns,
+                                true,
+                                (IcebergTableHandle) tableHandle)),
+                remainingExpressions.getDynamicFilterExpression());
+    }
+
+    @Override
+    protected TupleDomain<ColumnHandle> getUnenforcedConstraint(HivePartitionResult hivePartitionResult, Constraint<ColumnHandle> constraint)
+    {
+        return TupleDomain.withColumnDomains(constraint.getSummary().getDomains().get());
+    }
+
+    @Override
+    protected HivePartitionResult getHivePartitionResult(ConnectorMetadata metadata, ConnectorTableHandle tableHandle, Constraint<ColumnHandle> constraint, ConnectorSession session)
+    {
+        return null;
```

Comment:
Won't this cause an NPE in `BaseHiveFilterPushdown`?

Reply:
I think we should fail the query if the Iceberg table is partitioned and filter pushdown was turned on for now. @imjalpreet is there a way to get this information in this rule?